File: DSA/Manav-Rajput/Day11_question_2.java
import java.util.ArrayList;

class PriorityQueue {
    private ArrayList<Integer> heap;

    public PriorityQueue() {
        heap = new ArrayList<>();
    }

    public void insert(int value) {
        heap.add(value);
        int currentIndex = heap.size() - 1;

        // Bubble up
        while (currentIndex > 0) {
            int parentIndex = (currentIndex - 1) / 2;
            if (heap.get(currentIndex) < heap.get(parentIndex)) {
                // Swap current with parent
                int temp = heap.get(currentIndex);
                heap.set(currentIndex, heap.get(parentIndex));
                heap.set(parentIndex, temp);

                currentIndex = parentIndex;
            } else {
                break;
            }
        }
    }

    public int extract_min() {
        if (heap.isEmpty()) {
            throw new IllegalStateException("Priority queue is empty");
        }

        int minValue = heap.get(0);
        int lastValue = heap.remove(heap.size() - 1);

        if (!heap.isEmpty()) {
            heap.set(0, lastValue);
            int currentIndex = 0;

            // Bubble down
            while (true) {
                int leftChildIndex = 2 * currentIndex + 1;
                int rightChildIndex = 2 * currentIndex + 2;
                int smallestIndex = currentIndex;

                if (leftChildIndex < heap.size() && heap.get(leftChildIndex) < heap.get(smallestIndex)) {
                    smallestIndex = leftChildIndex;
                }

                if (rightChildIndex < heap.size() && heap.get(rightChildIndex) < heap.get(smallestIndex)) {
                    smallestIndex = rightChildIndex;
                }

                if (smallestIndex != currentIndex) {
                    // Swap current with smallest child
                    int temp = heap.get(currentIndex);
                    heap.set(currentIndex, heap.get(smallestIndex));
                    heap.set(smallestIndex, temp);

                    currentIndex = smallestIndex;
                } else {
                    break;
                }
            }
        }

        return minValue;
    }

    public static void main(String[] args) {
        PriorityQueue priorityQueue = new PriorityQueue();

        priorityQueue.insert(4);
        priorityQueue.insert(2);

        System.out.println(priorityQueue.extract_min()); // Output: 2
        System.out.println(priorityQueue.extract_min()); // Output: 4
    }
}

File: DSA/Ansh-Gulati/Day-11/Day-11_Question-2.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> heap;

void swap(int &a, int &b)
{
    int temp = a;
    a = b;
    b = temp;
}

void heapifyUp(int index)
{
    if (index == 0)
        return;
    int parentIndex = (index - 1) / 2;
    if (heap[parentIndex] > heap[index])
    {
        swap(heap[parentIndex], heap[index]);
        heapifyUp(parentIndex);
    }
}

void heapifyDown(int index)
{
    int leftChild = 2 * index + 1;
    int rightChild = 2 * index + 2;
    int smallest = index;

    if (leftChild < heap.size() && heap[leftChild] < heap[smallest])
    {
        smallest = leftChild;
    }
    if (rightChild < heap.size() && heap[rightChild] < heap[smallest])
    {
        smallest = rightChild;
    }
    if (smallest != index)
    {
        swap(heap[index], heap[smallest]);
        heapifyDown(smallest);
    }
}

void insert(int value)
{
    heap.push_back(value);
    heapifyUp(heap.size() - 1);
}

int extractMin()
{
    if (heap.empty())
    {
        cout << "Priority queue is empty" << endl;
        return -1;
    }
    int minValue = heap[0];
    heap[0] = heap.back();
    heap.pop_back();
    if (!heap.empty())
    {
        heapifyDown(0);
    }
    return minValue;
}

int main()
{
    insert(4);
    insert(2);
    cout << extractMin() << endl;
    cout << extractMin() << endl;

    return 0;
}